components/raftstore/src/store/fsm/apply.rs

```diff
@@ -370,6 +366,32 @@ impl ApplyContext {
         }
     }
 
+    /// Prepares WriteBatch.
+    ///
+    /// If `enable_multi_batch_write` was set false, we use first `RawWriteBatch`
+    /// of `RocksWriteBatchVec`. Otherwise we limit size of each `RawWriteBatch` in
+    /// `RocksWriteBatchVec` not exceed `WRITE_BATCH_LIMIT`.
+    pub fn prepare_write_batch(&mut self) {
+        if self.kv_wb.is_none() {
+            let kb_wb = {
+                if self.enable_multi_batch_write {
+                    self.engines
+                        .kv
+                        .c()
+                        .write_batch_vec(WRITE_BATCH_LIMIT, DEFAULT_APPLY_WB_SIZE)
+                } else {
+                    self.engines
+                        .kv
+                        .c()
+                        .write_batch_vec(0, DEFAULT_APPLY_WB_SIZE)
+                }
+            };
+            self.kv_wb = Some(kb_wb);
+            self.kv_wb_last_bytes = 0;
+            self.kv_wb_last_keys = 0;
+        }
```

Comment:
This looks to be the one place where the decision to use batched writes is made. It's a bit confusing to me and I think it could be simplified.

With the `supports_write_batch_vec` method, I'd expect that to indicate whether `WriteBatchVec` works at all. Instead what it means is "use a 0 write batch limit or use a constant write batch limit".

That could just be encoded in the abstraction itself. That is, leave all the parameters off of `write_batch_vec` and instead have it do the optimization if write batching is available and enabled, or fall back to not batching otherwise. That would be a more future-proof abstraction for other engines.

Reply:
Ok. I will replace it.

Reply:
The better solution is that we use trait to decide which kind of WriteBatch shall be created. So I hope to merged another PR at first: https://github.com/tikv/tikv/pull/7119